python/cudf_polars/cudf_polars/utils/config.py

```diff
@@ -5,209 +5,304 @@
-        try:
-            return self._hash_value
-        except AttributeError:
-            self._hash_value = hash(json.dumps(self.config_options))
-            return self._hash_value
-
-    @staticmethod
-    def validate(config: dict) -> dict:
-        """
-        Validate a configuration-option dictionary.
 
-        Parameters
-        ----------
-        config
-            GPUEngine configuration options to validate.
+@dataclasses.dataclass(frozen=True)
+class ParquetOptions:
+    """
+    Configuration for the cudf-polars Parquet engine.
+
+    Parameters
+    ----------
+    chunked
+        Whether to use cudf's ``ChunkedParquetReader`` to read the parquet
+        dataset in chunks. This is useful when reading very large parquet
+        files.
+    chunk_read_limit
+        Limit on total number of bytes to be returned per read, or 0 if
+        there is no limit.
+    pass_read_limit
+        Limit on the amount of memory used for reading and decompressing data
+        or 0 if there is no limit.
+    """
 
-        Returns
-        -------
-        Valid config-option dictionary.
+    chunked: bool = True
+    chunk_read_limit: int = 0
+    pass_read_limit: int = 0
 
-        Raises
-        ------
-        ValueError
-            If the configuration contains unsupported options.
-        """
-        if unsupported := (
-            config.keys()
-            - {"raise_on_fail", "parquet_options", "executor", "executor_options"}
-        ):
+    def __post_init__(self) -> None:
+        if not isinstance(self.chunked, bool):
+            raise TypeError("chunked must be a bool")
+        if not isinstance(self.chunk_read_limit, int):
+            raise TypeError("chunk_read_limit must be an int")
+        if not isinstance(self.pass_read_limit, int):
+            raise TypeError("pass_read_limit must be an int")
+
+
+@dataclasses.dataclass(frozen=True, eq=True)
+class StreamingExecutor:
+    """
+    Configuration for the cudf-polars streaming executor.
+
+    Parameters
+    ----------
+    scheduler
+        The scheduler to use for the streaming executor. ``Scheduler.SYNCHRONOUS``
+        by default.
+
+        Note ``scheduler="distributed"`` requires a Dask cluster to be running.
+
+    fallback_mode
+        How to handle errors when the GPU engine fails to execute a query.
+        ``FallbackMode.WARN`` by default.
+    max_rows_per_partition
+        The maximum number of rows to process per partition. 1_000_000 by default.
+        When the number of rows exceeds this value, the query will be split into
+        multiple partitions and executed in parallel.
+    cardinality_factor
+        A dictionary mapping column names floats strictly greater than 0 and less
+        than or equal to 1. Each factor estimates the fractional number of unique values
+        in the column. By default, a value of ``1.0`` is used for any column not
+        included in ``cardinality_factor``.
```

Comment:
`-ENOGRAMMAR`

BTW I would still love to not need this but it needs a big rethink of how we execute partitioned tasks, so...

Reply:
This is definitely a hack until we (hopefully) work out a better way to switch between tree- and shuffles-based algorithms.